### DISCUSSION TITLE
on_conflict_do_update SQLite have no effect on stmt

### DISCUSSION THREAD

**Discussion by iairsalem** [sqlalchemy/sqlalchemy#7565] [Usage Questions] [answered]
Hi, I have an issue with an insert statement (with many rows) that also has on_conflict_do_update.
I'm testing on SQLite.

Followed the docs:
https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#insert-on-conflict-upsert
code goes like this:
`

primary_key = User.__table__.c.id
stmt = insert(User.__table__).values(list_of_user_as_dicts) #option 1
stmt = insert(User.__table__) #option 2
update_dict = {c.name:c for c in stmt.excluded if not c.primary_key}
# update_dict gets the right data in.
stmt.on_conflict_do_update(index_elements=primary_key, set_=update_dict)
print(stmt) # doesn't print ON CONFLICT part...
session.execute(stmt) #option 1 #doesn't work
session.execute(stmt, list_of_user_as_dicts) #option 2 #doesn't work
`

Sample of printed stmt
`
INSERT INTO users (id, checksum, field1, field2, field3, field4, field5, field6, field7, field8, field9, field10, field11, field12, field13, field14, field15, field16, field17, field18, field19) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
-- about 40 rows per insert not to break the 999 limit. However, this should not bother the statement generator.
`

Hoping to get some enlightment from someone whe has already tested it.
Perhaps it's a bug?
Thanks.

**Comment by zzzeek** [comment, marked as answer]
> stmt.on_conflict_do_update(index_elements=primary_key, set_=update_dict) print(stmt) # doesn't print ON CONFLICT part... session.execute(stmt) #option 1 #doesn't work session.execute(stmt, list_of_user_as_dicts) #option 2 #doesn't work 

you need to assign the return value of "on_conflict_do_update()" to "stmt":

    stmt = stmt.on_conflict_do_update(...)

since code here is paraphrased that's all I can tell is the issue.

**Reply by iairsalem** [reply]
Thanks! totally overlooked that.
One more question, unrelated:
In your opinion, will it be faster to try to separate the INSERTS from the UPDATES (by performing additional SELECT queries and then dividing rows on python) or Just do the UPSERTS? (assume 10k rows, 50% new rows, 50% to-overwrite rows).
Usually that's a DB-specific question, but, given I'd have to execute more queries, perhaps it's better to do it in one query and let the DB handle the load (which is also pretty efficient at).

**Reply by zzzeek** [reply]
my understanding is upserts are pretty efficient if they do what you need.   I've never done a paid gig where the DB in use had them available, so I'm pretty good at running SELECT/INSERT/UPDATE in batches, but the straight upsert should always be faster than that because you no longer have round trips.
